File: Lab3/src/main/java/ua/lviv/iot/Manager/ComputerCompany.java
package ua.lviv.iot.Manager;

import ua.lviv.iot.Enums.Parts;

import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;

public class ComputerCompany {
    private List<Part> computerParts = new LinkedList<>();

    public ComputerCompany() {
    }

    public ComputerCompany(final List<Part> parts) {
        this.computerParts = parts;
    }

    public final void addPart(Part addPart) {
        computerParts.add(addPart);
    }

    public final List<Part> getComputerParts() {
        return computerParts;
    }

    public final List<Part> sortByPriceAndBrand(final List<Part> compParts) {
        compParts.sort(Comparator.comparing(Part::getBrand).thenComparing(Part::getPrice));
        return compParts;
    }


    public final List<Part> createPCFromPartsCheaperThan(final int maxPrice) {
        List<Part> cheapestParts = new LinkedList();
        cheapestParts.add(selectCheapestPart(maxPrice, Parts.CPU));
        cheapestParts.add(selectCheapestPart(maxPrice, Parts.MOTHERBOARD));
        cheapestParts.add(selectCheapestPart(maxPrice, Parts.VIDEOCARD));
        cheapestParts.add(selectCheapestPart(maxPrice, Parts.SCREEN));
        cheapestParts.add(selectCheapestPart(maxPrice, Parts.KEYBOARD));
        cheapestParts.add(selectCheapestPart(maxPrice, Parts.MOUSE));

        int total = 0;
        for (Part part : cheapestParts) {
            total += part.getPrice();
        }
        if (total <= maxPrice) {
            return sortByPriceAndBrand(cheapestParts);
        } else {
            return new LinkedList<Part>();
        }
    }

    public final Part selectCheapestPart(final int maxPrice, final Parts category) {
        Part min = new Part();
        min.setPrice(maxPrice);
        for (Part part : computerParts) {
            if (part.getCategory() == category && part.getPrice() < min.getPrice()) {
                min = part;
            }
        }
        return min;
    }
}

File: Lab12/ComputerCompany.py
from Part import Part
from Enums import PartType
from Enums import Parts

class ComputerCompany:
    computerParts = []

    def __init__(self, computerParts):
        self.computerParts = computerParts

    def addPart(self, addPart):
        self.computerParts.append(addPart)

    def sortByPrice(self):
        self.computerParts.sort(key=lambda x: x.price, reverse=False)
        self.printList()

    def selectCheapestPart(self, maxPrice, category):
        min = Part(maxPrice)
        for part in self.computerParts:
            if part.category == category and part.price < min.price:
                min = part
        return min

    def createPCFromPartsCheaperThan(self, maxPrice):
        cheapestParts = []
        cheapestParts.append(self.selectCheapestPart(maxPrice, Parts.CPU))
        cheapestParts.append(self.selectCheapestPart(maxPrice, Parts.MOTHERBOARD))
        cheapestParts.append(self.selectCheapestPart(maxPrice, Parts.VIDEOCARD))
        cheapestParts.append(self.selectCheapestPart(maxPrice, Parts.SCREEN))
        cheapestParts.append(self.selectCheapestPart(maxPrice, Parts.KEYBOARD))
        cheapestParts.append(self.selectCheapestPart(maxPrice, Parts.MOUSE))

        total = 0
        for part in cheapestParts:
            total += part.price

        if total <= maxPrice:
            for part in cheapestParts:
                print(part)
        else:
            return self.computerParts

    def printList(self):
        for i in self.computerParts:
            print(i)


